src/sysupdate/sysupdated.c

```diff
@@ -0,0 +1,1911 @@
+                return 0;
+
+        *ret = j;
+        return 1;
+}
+
+static int job_node_enumerator(
+                sd_bus *bus,
+                const char *path,
+                void *userdata,
+                char ***nodes,
+                sd_bus_error *error) {
+
+        _cleanup_strv_free_ char **l = NULL;
+        Manager *m = ASSERT_PTR(userdata);
+        Job *j;
+        unsigned k = 0;
+
+        l = new0(char*, hashmap_size(m->jobs) + 1);
+        if (!l)
+                return -ENOMEM;
+
+        HASHMAP_FOREACH(j, m->jobs) {
+                l[k] = strdup(j->object_path);
+                if (!l[k])
+                        return -ENOMEM;
+                k++;
+        }
+
+        *nodes = TAKE_PTR(l);
+        return 1;
+}
+
+static const sd_bus_vtable job_vtable[] = {
+        SD_BUS_VTABLE_START(0),
+
+        SD_BUS_PROPERTY("Id", "u", NULL, offsetof(Job, id), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Type", "s", job_property_get_type, offsetof(Job, type), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Offline", "b", NULL, offsetof(Job, offline), SD_BUS_VTABLE_PROPERTY_CONST),
+        SD_BUS_PROPERTY("Progress", "u", NULL, offsetof(Job, progress_percent), SD_BUS_VTABLE_PROPERTY_EMITS_CHANGE),
+
+        SD_BUS_METHOD("Cancel", NULL, NULL, job_method_cancel, SD_BUS_VTABLE_UNPRIVILEGED),
+
+        SD_BUS_VTABLE_END
+};
+
+static const BusObjectImplementation job_object = {
+        "/org/freedesktop/sysupdate1/job",
+        "org.freedesktop.sysupdate1.Job",
+        .fallback_vtables = BUS_FALLBACK_VTABLES({job_vtable, job_object_find}),
+        .node_enumerator = job_node_enumerator,
+};
+
+static Target *target_free(Target *t) {
+        if (!t)
+                return NULL;
+
+        if (t->manager)
+                hashmap_remove(t->manager->targets, t->id);
+
+        free(t->name);
+        free(t->path);
+        free(t->version);
+        free(t->appstream);
+        free(t->id);
+
+        return mfree(t);
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC(Target*, target_free);
+DEFINE_HASH_OPS_WITH_VALUE_DESTRUCTOR(target_hash_ops, char, string_hash_func, string_compare_func,
+                                      Target, target_free);
+
+static int target_new(Manager *m, TargetClass class, const char *name, const char *path, Target **ret) {
+        _cleanup_(target_freep) Target *t = NULL;
+        int r;
+
+        assert(m);
+        assert(ret);
+
+        t = new(Target, 1);
+        if (!t)
+                return -ENOMEM;
+
+        *t = (Target) {
+                .manager = m,
+                .class = class,
+                .image_type = _IMAGE_TYPE_INVALID,
+        };
+
+        t->name = strdup(name);
+        if (!t->name)
+                return -ENOMEM;
+
+        t->path = strdup(path);
+        if (!t->path)
+                return -ENOMEM;
+
+        if (class == TARGET_HOST)
+                t->id = strdup("host"); /* This is what appears in the object path */
```

Comment:
so in machinectl and image discovery we usually used the special name ".host" for this.